sound/drivers/pimidi.c

```diff
@@ -0,0 +1,1119 @@
+
+	INIT_WORK(&instance->drdy_handler, pimidi_drdy_handler);
+	mutex_init(&instance->comm_lock);
+
+	err = devm_request_irq(&client->dev, client->irq, pimidi_drdy_interrupt_handler,
+			       IRQF_SHARED | IRQF_TRIGGER_LOW, "data_ready_int", instance);
+
+	if (err != 0) {
+		printe(instance, "data_available IRQ request failed! %d", err);
+		goto finalize;
+	}
+
+	irq_initialized = true;
+
+	err = pimidi_read_version(&instance->version, instance);
+	if (err < 0) {
+		printe(instance, "pimidi_read_version failed: %d", err);
+		goto finalize;
+	}
+
+	err = pimidi_read_serial(instance->serial, instance);
+	if (err < 0) {
+		printe(instance, "pimidi_read_serial failed: %d", err);
+		goto finalize;
+	} else if (instance->serial[0] != 'P' || instance->serial[1] != 'M' ||
+		   strlen(instance->serial) != 10) {
+		printe(instance, "Unexpected serial number: %s", instance->serial);
+		err = -EIO;
+		goto finalize;
+	}
+
+	printi(instance, "pimidi%d hw:%d version %u.%u.%u-%u, serial %s",
+	       d,
+	       card->number,
+	       instance->version.major,
+	       instance->version.minor,
+	       instance->version.build,
+	       instance->version.hwrev,
+	       instance->serial
+	       );
+
+	strscpy(card->driver, "snd-pimidi", sizeof(card->driver));
+	snprintf(card->shortname, sizeof(card->shortname), "pimidi%d", d);
+	snprintf(card->longname, sizeof(card->longname), "pimidi%d %s", d, instance->serial);
+	snprintf(card->id, sizeof(card->id), "pimidi%d", d);
+
+	snprintf(pimidi_find_substream(rawmidi, SNDRV_RAWMIDI_STREAM_OUTPUT, 0)->name,
+		 10u, "pimidi%d-a", d);
+	snprintf(pimidi_find_substream(rawmidi, SNDRV_RAWMIDI_STREAM_INPUT,  0)->name,
+		 10u, "pimidi%d-a", d);
+	snprintf(pimidi_find_substream(rawmidi, SNDRV_RAWMIDI_STREAM_OUTPUT, 1)->name,
+		 10u, "pimidi%d-b", d);
+	snprintf(pimidi_find_substream(rawmidi, SNDRV_RAWMIDI_STREAM_INPUT,  1)->name,
+		 10u, "pimidi%d-b", d);
+
+	err = snd_card_ro_proc_new(card, "a-tx", &instance->midi_port[0].tx_cnt,
+				   pimidi_proc_stat_show);
+	err = snd_card_ro_proc_new(card, "a-rx", &instance->midi_port[0].rx_cnt,
+				   pimidi_proc_stat_show);
+	err = snd_card_ro_proc_new(card, "b-tx", &instance->midi_port[1].tx_cnt,
+				   pimidi_proc_stat_show);
+	err = snd_card_ro_proc_new(card, "b-rx", &instance->midi_port[1].rx_cnt,
+				   pimidi_proc_stat_show);
+	err = snd_card_ro_proc_new(card, "serial", instance, pimidi_proc_serial_show);
+	err = snd_card_ro_proc_new(card, "version", instance, pimidi_proc_version_show);
+	err = snd_card_ro_proc_new(card, "hwrev", instance, pimidi_proc_hwrev_show);
+	if (err < 0) {
+		printe(instance, "snd_card_ro_proc_new failed: %d", err);
+		goto finalize;
+	}
+
+	err = snd_card_register(card);
+	if (err < 0) {
+		printe(instance, "snd_card_register failed: %d", err);
+		goto finalize;
+	}
+
+finalize:
+	if (err) {
+		instance->stopping = true;
+		cancel_work_sync(&instance->drdy_handler);
+		if (irq_initialized)
+			devm_free_irq(&client->dev, client->irq, instance);
```

Comment:
It should not be necessary to call `devm_free_irq` unless you are intended to release the IRQ before the device is released. Returning an error from the probe method is not one of those situations, so you can delete this line (along with all mentions of `irq_initialized`).

Reply:
You are right, thank you, I've removed the lines in the latest push.